### PR TITLE
Implement verify

### PR DESCRIPTION
**Pull request by HarshitaKalani**
Continuing https://github.com/lfortran/lfortran/pull/3497

### CODE REVIEW

File: b.f90 [hunk 1]
```diff
@@ -0,0 +1,6 @@
+program main
+    character(7) :: a, b
+    a = "fortran"
+    b = "foo"
+    print*, verify(a, b)
+end
```
Comment (HarshitaKalani) [inline b.f90:6]: This example throws the following error:
```console
$ lfortran b.f90
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  Binary file "/home/harshita/Desktop/lfortran/src/bin/lfortran", in _start()
  File "./csu/../csu/libc-start.c", line 392, in __libc_start_main_impl()
  File "./csu/../sysdeps/nptl/libc_start_call_main.h", line 58, in __libc_start_call_main()
  File "/home/harshita/Desktop/lfortran/src/bin/lfortran.cpp", line 2394, in ??
    return main_app(argc, argv);
  File "/home/harshita/Desktop/lfortran/src/bin/lfortran.cpp", line 2355, in main_app(int, char**)
    err = compile_to_object_file(arg_file, tmp_o, false,
  File "/home/harshita/Desktop/lfortran/src/bin/lfortran.cpp", line 949, in ??
    res = fe.get_llvm3(*asr, lpm, diagnostics, infile);
  File "/home/harshita/Desktop/lfortran/src/lfortran/fortran_evaluator.cpp", line 360, in LCompilers::FortranEvaluator::get_llvm3(LCompilers::ASR::TranslationUnit_t&, LCompilers::PassManager&, LCompilers::diag::Diagnostics&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    compiler_options, run_fn, infile);
  File "/home/harshita/Desktop/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 9745, in LCompilers::asr_to_llvm(LCompilers::ASR::TranslationUnit_t&, LCompilers::diag::Diagnostics&, llvm::LLVMContext&, Allocator&, LCompilers::PassManager&, LCompilers::CompilerOptions&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    v.visit_asr((ASR::asr_t&)asr);
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/asr.h", line 5136, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_asr(LCompilers::ASR::asr_t const&)
    void visit_asr(const asr_t &b) { visit_asr_t(b, self()); }
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/asr.h", line 5112, in ??
    case asrType::unit: { v.visit_unit((const unit_t &)x); return; }
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/asr.h", line 5137, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_unit(LCompilers::ASR::unit_t const&)
    void visit_unit(const unit_t &b) { visit_unit_t(b, self()); }
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/asr.h", line 4842, in ??
    case unitType::TranslationUnit: { v.visit_TranslationUnit((const TranslationUnit_t &)x); return; }
  File "/home/harshita/Desktop/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 918, in LCompilers::ASRToLLVMVisitor::visit_TranslationUnit(LCompilers::ASR::TranslationUnit_t const&)
    visit_Function(*ASR::down_cast<ASR::Function_t>(item.second));
  File "/home/harshita/Desktop/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 3726, in LCompilers::ASRToLLVMVisitor::visit_Function(LCompilers::ASR::Function_t const&)
    instantiate_function(x);
  File "/home/harshita/Desktop/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 3766, in LCompilers::ASRToLLVMVisitor::instantiate_function(LCompilers::ASR::Function_t const&)
    llvm::FunctionType* function_type = llvm_utils->get_function_type(x, module.get());
  File "/home/harshita/Desktop/lfortran/src/libasr/codegen/llvm_utils.cpp", line 955, in LCompilers::LLVMUtils::get_function_type(LCompilers::ASR::Function_t const&, llvm::Module*)
    ASR::ttype_t *return_var_type0 = ASRUtils::EXPR2VAR(x.m_return_var)->m_type;
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/asr_utils.h", line 284, in ??
    ASR::down_cast<ASR::Var_t>(f)->m_v));
AssertFailed: is_a<T>(*f)
```

Comment (HarshitaKalani) [reply b.f90]:
```console
$ lfortran a.f90
; ModuleID = 'LFortran'
source_filename = "LFortran"

@0 = private unnamed_addr constant [8 x i8] c"fortran\00", align 1
@1 = private unnamed_addr constant [4 x i8] c"foo\00", align 1
@2 = private unnamed_addr constant [2 x i8] c" \00", align 1
@3 = private unnamed_addr constant [2 x i8] c"\0A\00", align 1
@4 = private unnamed_addr constant [5 x i8] c"%d%s\00", align 1

define i32 @_lcompilers_verify_str(i8** %x, i8** %y, i1* %z, i32* %k) {
.entry:
  %_lcompilers_verify_str = alloca i32, align 4
  %i = alloca i32, align 4
  %j = alloca i32, align 4
  %matched = alloca i1, align 1
  store i32 0, i32* %j, align 4
  %0 = load i1, i1* %z, align 1
  %1 = zext i1 %0 to i32
  %2 = icmp eq i32 %1, 1
  br i1 %2, label %then, label %else8

then:                                             ; preds = %.entry
  %3 = load i8*, i8** %x, align 8
  %4 = alloca i8*, align 8
  store i8* %3, i8** %4, align 8
  %5 = call i32 @_lfortran_str_len(i8** %4)
  store i32 %5, i32* %i, align 4
  br label %loop.head

loop.head:                                        ; preds = %ifcont6, %then
  %6 = load i32, i32* %i, align 4
  %7 = icmp sge i32 %6, 1
  br i1 %7, label %loop.body, label %loop.end7

loop.body:                                        ; preds = %loop.head
  store i1 false, i1* %matched, align 1
  store i32 1, i32* %j, align 4
  br label %loop.head1

loop.head1:                                       ; preds = %ifcont, %loop.body
  %8 = load i32, i32* %j, align 4
  %9 = load i8*, i8** %y, align 8
  %10 = alloca i8*, align 8
  store i8* %9, i8** %10, align 8
  %11 = call i32 @_lfortran_str_len(i8** %10)
  %12 = icmp sle i32 %8, %11
  br i1 %12, label %loop.body2, label %loop.end

loop.body2:                                       ; preds = %loop.head1
  %13 = load i8*, i8** %x, align 8
  %14 = load i32, i32* %i, align 4
  %15 = sub i32 %14, 1
  %16 = load i32, i32* %i, align 4
  %17 = call i8* @_lfortran_str_slice(i8* %13, i32 %15, i32 %16, i32 1, i1 true, i1 true)
  %18 = load i8*, i8** %y, align 8
  %19 = load i32, i32* %j, align 4
  %20 = sub i32 %19, 1
  %21 = load i32, i32* %j, align 4
  %22 = call i8* @_lfortran_str_slice(i8* %18, i32 %20, i32 %21, i32 1, i1 true, i1 true)
  %23 = zext i8* %17 to i32
  %24 = zext i8* %22 to i32
  %25 = icmp eq i32 %23, %24
  br i1 %25, label %then3, label %else

then3:                                            ; preds = %loop.body2
  store i1 true, i1* %matched, align 1
  br label %ifcont

else:                                             ; preds = %loop.body2
  br label %ifcont

ifcont:                                           ; preds = %else, %then3
  %26 = load i32, i32* %j, align 4
  %27 = add i32 %26, 1
  store i32 %27, i32* %j, align 4
  br label %loop.head1

loop.end:                                         ; preds = %loop.head1
  %28 = load i1, i1* %matched, align 1
  %29 = zext i1 %28 to i32
  %30 = icmp eq i32 %29, 0
  br i1 %30, label %then4, label %else5

then4:                                            ; preds = %loop.end
  %31 = load i32, i32* %i, align 4
  store i32 %31, i32* %_lcompilers_verify_str, align 4
  br label %loop.end7

unreachable_after_exit:                           ; No predecessors!
  br label %ifcont6

else5:                                            ; preds = %loop.end
  br label %ifcont6

ifcont6:                                          ; preds = %else5, %unreachable_after_exit
  %32 = load i32, i32* %i, align 4
  %33 = sub i32 %32, 1
  store i32 %33, i32* %i, align 4
  br label %loop.head

loop.end7:                                        ; preds = %then4, %loop.head
  br label %ifcont22

else8:                                            ; preds = %.entry
  store i32 1, i32* %i, align 4
  br label %loop.head9

loop.head9:                                       ; preds = %ifcont20, %else8
  %34 = load i32, i32* %i, align 4
  %35 = load i8*, i8** %x, align 8
  %36 = alloca i8*, align 8
  store i8* %35, i8** %36, align 8
  %37 = call i32 @_lfortran_str_len(i8** %36)
  %38 = icmp sle i32 %34, %37
  br i1 %38, label %loop.body10, label %loop.end21

loop.body10:                                      ; preds = %loop.head9
  store i1 false, i1* %matched, align 1
  store i32 1, i32* %j, align 4
  br label %loop.head11

loop.head11:                                      ; preds = %ifcont15, %loop.body10
  %39 = load i32, i32* %j, align 4
  %40 = load i8*, i8** %y, align 8
  %41 = alloca i8*, align 8
  store i8* %40, i8** %41, align 8
  %42 = call i32 @_lfortran_str_len(i8** %41)
  %43 = icmp sle i32 %39, %42
  br i1 %43, label %loop.body12, label %loop.end16

loop.body12:                                      ; preds = %loop.head11
  %44 = load i8*, i8** %x, align 8
  %45 = load i32, i32* %i, align 4
  %46 = sub i32 %45, 1
  %47 = load i32, i32* %i, align 4
  %48 = call i8* @_lfortran_str_slice(i8* %44, i32 %46, i32 %47, i32 1, i1 true, i1 true)
  %49 = load i8*, i8** %y, align 8
  %50 = load i32, i32* %j, align 4
  %51 = sub i32 %50, 1
  %52 = load i32, i32* %j, align 4
  %53 = call i8* @_lfortran_str_slice(i8* %49, i32 %51, i32 %52, i32 1, i1 true, i1 true)
  %54 = zext i8* %48 to i32
  %55 = zext i8* %53 to i32
  %56 = icmp eq i32 %54, %55
  br i1 %56, label %then13, label %else14

then13:                                           ; preds = %loop.body12
  store i1 true, i1* %matched, align 1
  br label %ifcont15

else14:                                           ; preds = %loop.body12
  br label %ifcont15

ifcont15:                                         ; preds = %else14, %then13
  %57 = load i32, i32* %j, align 4
  %58 = add i32 %57, 1
  store i32 %58, i32* %j, align 4
  br label %loop.head11

loop.end16:                                       ; preds = %loop.head11
  %59 = load i1, i1* %matched, align 1
  %60 = zext i1 %59 to i32
  %61 = icmp eq i32 %60, 0
  br i1 %61, label %then17, label %else19

then17:                                           ; preds = %loop.end16
  %62 = load i32, i32* %i, align 4
  store i32 %62, i32* %_lcompilers_verify_str, align 4
  br label %loop.end21

unreachable_after_exit18:                         ; No predecessors!
  br label %ifcont20

else19:                                           ; preds = %loop.end16
  br label %ifcont20

ifcont20:                                         ; preds = %else19, %unreachable_after_exit18
  %63 = load i32, i32* %i, align 4
  %64 = add i32 %63, 1
  store i32 %64, i32* %i, align 4
  br label %loop.head9

loop.end21:                                       ; preds = %then17, %loop.head9
  br label %ifcont22

ifcont22:                                         ; preds = %loop.end21, %loop.end7
  br label %return

return:                                           ; preds = %ifcont22
  %65 = load i32, i32* %_lcompilers_verify_str, align 4
  ret i32 %65
}

declare i32 @_lfortran_str_len(i8**)

declare i8* @_lfortran_str_slice(i8*, i32, i32, i32, i1, i1)

define i32 @main(i32 %0, i8** %1) {
.entry:
  %call_arg_value1 = alloca i32, align 4
  %call_arg_value = alloca i1, align 1
  call void @_lpython_set_argv(i32 %0, i8** %1)
  %a = alloca i8*, align 8
  %2 = call i8* (i32, ...) @_lfortran_malloc(i32 2)
  call void (i32, i8*, ...) @_lfortran_string_init(i32 2, i8* %2)
  store i8* %2, i8** %a, align 8
  %3 = load i8*, i8** %a, align 8
  %b = alloca i8*, align 8
  %4 = call i8* (i32, ...) @_lfortran_malloc(i32 2)
  call void (i32, i8*, ...) @_lfortran_string_init(i32 2, i8* %4)
  store i8* %4, i8** %b, align 8
  %5 = load i8*, i8** %b, align 8
  call void @_lfortran_strcpy(i8** %a, i8* getelementptr inbounds ([8 x i8], [8 x i8]* @0, i32 0, i32 0), i8 0)
  call void @_lfortran_strcpy(i8** %b, i8* getelementptr inbounds ([4 x i8], [4 x i8]* @1, i32 0, i32 0), i8 0)
  store i1 true, i1* %call_arg_value, align 1
  store i32 8, i32* %call_arg_value1, align 4
  %6 = call i32 @_lcompilers_verify_str(i8** %a, i8** %b, i1* %call_arg_value, i32* %call_arg_value1)
  call void (i8*, ...) @_lfortran_printf(i8* getelementptr inbounds ([5 x i8], [5 x i8]* @4, i32 0, i32 0), i32 %6, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @3, i32 0, i32 0))
  ret i32 0
}

declare void @_lpython_set_argv(i32, i8**)

declare i8* @_lfortran_malloc(i32, ...)

declare void @_lfortran_string_init(i32, i8*, ...)

declare void @_lfortran_strcpy(i8**, i8*, i8)

declare void @_lfortran_printf(i8*, ...)
code generation error: asr_to_llvm: module failed verification. Error:
ZExt only operates on integer
  %23 = zext i8* %17 to i32
ZExt only operates on integer
  %24 = zext i8* %22 to i32
ZExt only operates on integer
  %54 = zext i8* %48 to i32
ZExt only operates on integer
  %55 = zext i8* %53 to i32



Note: Please report unclear or confusing messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 2]
```diff
@@ -534,6 +537,8 @@ namespace IntrinsicElementalFunctionRegistry {
             "rrspacing"},
         {static_cast<int64_t>(IntrinsicElementalFunctions::Repeat),
             "Repeat"},
+        {static_cast<int64_t>(IntrinsicElementalFunctions::StringContainsSet),
+            "Verify"},
```
Comment (certik) [inline src/libasr/pass/intrinsic_function_registry.h:541]:
```suggestion
            "StringContainsSet"},
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 3]
```diff
@@ -721,6 +726,7 @@ namespace IntrinsicElementalFunctionRegistry {
                 {"digits", {&Digits::create_Digits, &Digits::eval_Digits}},
                 {"rrspacing", {&Rrspacing::create_Rrspacing, &Rrspacing::eval_Rrspacing}},
                 {"repeat", {&Repeat::create_Repeat, &Repeat::eval_Repeat}},
+                {"verify", {&StringContainsSet::create_StringContainsSet, &StringContainsSet::eval_StringContainsSet}},
```
Comment (certik) [inline src/libasr/pass/intrinsic_function_registry.h:729]:
```suggestion
                {"string_contains_set", {&StringContainsSet::create_StringContainsSet, &StringContainsSet::eval_StringContainsSet}},
```

File: src/libasr/pass/intrinsic_function_registry.h (hunk 2 above)
Comment (certik) [reply src/libasr/pass/intrinsic_function_registry.h:541]: If this is too complicated right now, just leave it and we can fix it in a separate PR.

File: src/libasr/pass/intrinsic_functions.h [hunk 4]
```diff
@@ -3562,6 +3563,143 @@ namespace Repeat {
...
+        }, {
+            b.Assignment(i, b.i(1, return_type)),
+            b.While(iLtE(i, str_len), while_loop_body_else)
+        }));
+
+        ASR::symbol_t *f_sym = make_ASR_Function_t(fn_name, fn_symtab, dep, args,
+            body, result, ASR::abiType::Source, ASR::deftypeType::Implementation, nullptr);
+        scope->add_symbol(fn_name, f_sym);
+        return b.Call(f_sym, new_args, return_type, nullptr);
+    }
```
Comment (certik) [inline src/libasr/pass/intrinsic_functions.h:3699]: Just keep removing things and compiling to LLVM until it works. Use "bisection" to figure out which part breaks.

**Comment by HarshitaKalani** [comment]
Fails due to: https://github.com/lfortran/lfortran/issues/3584

**Review by certik — APPROVED** [review]
I think that this is good to merge, thanks @HarshitaKalani !

**Pull request merged**